finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4DispatcherBase.scala

```diff
@@ -104,6 +139,11 @@ trait Netty4DispatcherBase[SendMsg <: Message, RecvMsg <: Message] {
     transport.read().rescue(wrapRemoteEx)
 
   protected[this] def demux(): Future[Unit] = {
+    detector.onClose.ensure {
+      log.debug(s"[%s] did not receive HTTP/2 ping ack from remote [%s]", prefix, transport.context.remoteAddress)
+      goAway(GoAway.InternalError); ()
```

Comment:
I'm looking at the mux implementation for comparison and it doesn't explicitly tear down the connection on close of the detector.  Instead, it just sets its status to the detector's status.

I guess some other layer of the stack closes the service if the status becomes closed, maybe?

Reply:
Testing this out with the SIGSTOP scenario we discussed. Without this, the client side stops pinging but the connection is still `ESTABLISHED`. I am not really sure if there is some kind of cleanup layer in the stack that detects if a service is closed. But I think we may have to tear down the connection manually.

Reply:
My understanding is that failure detection is interacts with the load balancer & probation, so that "failed" nodes are not considered for new requests.

These pings _can_ cause the _operating system_ to notify the process that a closed connection was actually closed -- i.e. if writing to the socket fails. (Otherwise, we may not try to write to the socket and wouldn't have cause to be notified of the disconnect).

SIGSTOPing a process for just a short time probably wouldn't be enough to cause the OS to tell the process the connection has closed. It should be sufficient to see the load balancer put the endpoint in probation; though this would only impact the LB, really, if there are more than one endpoint to choose from.